### ISSUE TITLE
SuccessResultOf matcher for synchronous Deferred testing

### ISSUE THREAD

**Issue by mithrandi** [testing-cabal/testtools#93]
I wrote this, based on `SynchronousTestCase.successResultOf` from Twisted:

```
class SuccessResultOf(object):
    """
    Match if a Deferred has fired with a result.
    """
    def __init__(self, resultMatcher=None):
        self.resultMatcher = resultMatcher


    def __str__(self):
        return 'SuccessResultOf(%s)' % (self.resultMatcher or '')


    def match(self, matchee):
        result = []
        matchee.addBoth(result.append)
        if not result:
            return Mismatch(
                'Success result expected on %r, found no result instead' % (
                    matchee,))
        elif isinstance(result[0], Failure):
            return Mismatch(
                'Success result expected on %r,'
                ' found failure result instead:\n%s' % (
                    matchee, result[0].getTraceback()))
        elif self.resultMatcher:
            return self.resultMatcher.match(result[0])
        return None
```

Presumably it would make sense to add FailureResultOf and NoResult matchers as well; the implementation of these would be a trivial permutation of the SuccessResultOf logic.

**Comment by rbtcollins** [comment]
ping?

**Comment by mithrandi** [comment]
Sorry, life happened ;) I haven't completely forgotten about this ticket, just got bumped a bit down my todo list.

**Comment by jml** [comment]
FWIW, I'm working on this: https://github.com/jml/testtools/tree/deferred-matchers

**Comment by mithrandi** [comment]
Merged in a5ea4b5365b8a2426e298be6b5182cfa529905fc.
